receiver/m365receiver/logs.go

```diff
@@ -0,0 +1,537 @@
+					l.logger.Error("error while polling for logs", zap.Error(err))
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// spins a go routine for each audit type/endpoint
+func (l *m365LogsReceiver) pollLogs(ctx context.Context) error {
+	var st string
+	if l.record.NextStartTime != nil {
+		st = l.record.NextStartTime.Format(layout)
+	} else {
+		st = pcommon.NewTimestampFromTime(time.Now().Add(-l.pollInterval)).AsTime().UTC().Format(layout)
+	}
+
+	now := time.Now().UTC()
+
+	auditWG := &sync.WaitGroup{}
+	auditWG.Add(5)
+	for i := 0; i < len(l.audits); i++ {
+		go l.poll(ctx, now, st, &l.audits[i], auditWG)
+	}
+	auditWG.Wait()
+
+	l.record.NextStartTime = &now
+	return l.checkpoint(ctx)
+}
+
+// collects log data from endpoint, transforms logs, consumes logs
+func (l *m365LogsReceiver) poll(ctx context.Context, now time.Time, st string, audit *auditMetaData, wg *sync.WaitGroup) {
+	defer wg.Done()
+	if !audit.enabled {
+		return
+	}
+
+	l.mu.Lock()
+	data, err := l.client.GetJSON(ctx, l.root+audit.route, now.Format(layout), st)
+	if err != nil {
+		if err.Error() == "authorization denied" { // troubleshoot stale token
+			l.logger.Debug("possible stale token; attempting to regenerate")
+			err = l.client.GetToken()
+			if err != nil { // something went wrong generating token
+				l.logger.Error("error creating authorization token", zap.Error(err))
+				l.mu.Unlock()
+				return
+			}
+			data, err = l.client.GetJSON(ctx, l.root+audit.route, now.Format(layout), st)
+			if err != nil { // not a stale token error, unsure what is wrong
+				l.logger.Error("unable to retrieve logs", zap.Error(err))
+				l.mu.Unlock()
+				return
+			}
+		} else {
+			l.logger.Error("error retrieving logs", zap.Error(err))
+			l.logger.Error(audit.route) //debug only
+			l.mu.Unlock()
+			return
+		}
+	}
+	l.mu.Unlock()
+
+	logs := l.transformLogs(pcommon.NewTimestampFromTime(now), audit, data)
+
+	l.consumerMu.Lock()
+	if logs.LogRecordCount() > 0 {
+		if err = l.consumer.ConsumeLogs(ctx, logs); err != nil {
+			l.logger.Error("error consuming events", zap.Error(err))
+		}
+	}
+	l.consumerMu.Unlock()
+}
+
+// constructs logs from logData
+func (l *m365LogsReceiver) transformLogs(now pcommon.Timestamp, audit *auditMetaData, data logData) plog.Logs {
+	logs := plog.NewLogs()
+	resourceLogs := logs.ResourceLogs().AppendEmpty()
+	ra := resourceLogs.Resource().Attributes()
+	ra.PutStr("m365.audit", audit.name)
+	ra.PutStr("m365.organization_id", l.cfg.TenantID)
+
+	for i := 0; i < len(data.logs); i++ {
+		log := data.logs[i]
+		logRecord := resourceLogs.ScopeLogs().AppendEmpty().LogRecords().AppendEmpty()
+
+		// log body
+		logRecord.Body().SetStr("\"C" + data.body[i])
+
+		// timestamp
+		ts, err := time.Parse(layout, log.CreationTime)
+		if err != nil {
+			l.logger.Warn("unable to interpret when an event was created, expecting a RFC3339 timestamp", zap.String("timestamp", log.CreationTime), zap.String("log", log.ID))
```

Comment:
Is there a reason we're not using time.RFC3339 as the layout?

Reply:
the timestamps returned by the api don't include the time zone offsets that are specified in time.RFC3339, so the time.Parse call was always returning a non nil error and setting the timestamp to time.Now instead of log.CreationTime, so I made a custom layout to better match CreationTime